src/osd/OSD.cc

```diff
@@ -728,39 +717,80 @@ void OSDService::check_nearfull_warning(const osd_stat_t &osd_stat)
   // much space is available to use than how much we've already used.
   float ratio = ((float)(osd_stat.kb - osd_stat.kb_avail)) /
     ((float)osd_stat.kb);
-  float nearfull_ratio = get_failsafe_nearfull_ratio();
-  float full_ratio = get_failsafe_full_ratio();
   cur_ratio = ratio;
 
-  if (full_ratio > 0 && ratio > full_ratio) {
-    new_state = FULL;
-  } else if (nearfull_ratio > 0 && ratio > nearfull_ratio) {
-    new_state = NEAR;
-  } else {
+  // The OSDMap ratios take precendence.  So if the failsafe is .95 and
+  // the admin sets the cluster full to .96, the failsafe moves up to .96
+  // too.  (Not that having failsafe == full is ideal, but it's better than
+  // dropping writes before the clusters appears full.)
+  OSDMapRef osdmap = get_osdmap();
+  if (!osdmap || osdmap->get_epoch() == 0) {
     cur_state = NONE;
     return;
   }
+  float nearfull_ratio = osdmap->get_nearfull_ratio();
+  float full_ratio = std::max(osdmap->get_full_ratio(), nearfull_ratio);
+  float failsafe_ratio = std::max(get_failsafe_full_ratio(), full_ratio);
 
+  if (full_ratio <= 0 ||
+      nearfull_ratio <= 0) {
+    derr << __func__ << " full_ratio or nearfull_ratio is <= 0" << dendl;
+    cur_state = NONE;
+    return;
+  }
+
+  enum s_names new_state;
+  if (ratio > failsafe_ratio) {
+    new_state = FAILSAFE;
+  } else if (ratio > full_ratio) {
+    new_state = FULL;
+  } else if (ratio > nearfull_ratio) {
+    new_state = NEARFULL;
+  } else {
+    new_state = NONE;
+  }
+  dout(20) << __func__ << " cur ratio " << ratio
+	   << ". nearfull_ratio " << nearfull_ratio
+	   << ", full_ratio " << full_ratio
+	   << ", failsafe_ratio " << failsafe_ratio
+	   << ", new state " << get_full_state_name(new_state)
+	   << dendl;
+
+  // warn
   if (cur_state != new_state) {
+    dout(10) << __func__ << " " << get_full_state_name(cur_state)
+	     << " -> " << get_full_state_name(new_state) << dendl;
+    if (new_state == FAILSAFE) {
+      clog->error() << "failsafe engaged, dropping updates, now "
+		    << (int)(ratio * 100) << "% full";
```

Comment:
My merged change uses  (int)roundf(ratio * 100)  which is being dropped in 2 places by your change